### DISCUSSION TITLE
Potential Web Worker use

### DISCUSSION THREAD

**Discussion by rhurlbatt** [anvilistas/anvil-labs#81] [Q&A]
Not that I know anything about web workers but I think I have a use case and would like some guidance on whether I am barking up the right tree or this is in fact a fire hydrant.

I have an app that carries out inspections, large inspections. typically 100 odd questions with 50 or 60 photos to be attached which are temporarily saved to IndexedDB before upload.

The problem we have faced is that uploading that volume of data needs to be done in chunks to avoid time outs and this locks up the app while upload takes place.

Could I use a web worker to handle this in the background leaving the user free to move on to other tasks?

Could I also use the web worker to check whether there is an internet connection before proceeding?

Thanks in advance.

Rick.

**Comment by s-cork** [comment]
Could you create the most minimal example?
Something that wouldn't need a web worker
Then I can see if I can modify it to use the web-worker api
I think it might be possible

The main limitation is that we only have access to the stdlib and not any anvil module
We do however have access to the window object (or rather the web-worker instance) so we might be able to do some clever things with bytes, wrapping the indexedDB interface, and doing an api request instead of anvil.server.call

It might not work but it'll be fun trying

**Comment by rhurlbatt** [comment]
Yes I sure can make a clone. I was about to start building a test app for this anyway so I will do that with the idea to clone that here.
Thanks Stu.

**Comment by rhurlbatt** [comment]
Ok so I have created a minimal example.

The basic premise is that to upload lots of files/images we need a consistent internet connection. If there is no connection when the files are uploaded they should be stored locally until they can be uploaded.

The example takes files from a `file_loader` and stores them in Indexed DB when the `Save to IDB` button is pressed. This could be linked to the `file_loader` change event but it will need to handle deletions if files are unselected in the `file_loader` When you press the `Sync to Cloud` button, a function gets client writable rows and uploads the file from Indexed DB to the datatable.

If files are large or there are lots it takes a bit of time (but it doesn't time out since it is using client writable views).

If this could be off loaded to a background process via a web worker then the main app can continue relatively unhindered.

Ultimately it would be good to extend this functionality to entire table rows but for now being able to upload the files separately would be sufficient.

I would be keen to make this work to include in Anvil Extras - something like this would have made my life a whole lot easier 2 years ago 

[Clone link](https://anvil.works/build#clone:MSO4IUIKGHP6FJ6D=ZLRDOKAMBYZXZODFRKO4UAE6)

[Example Link](https://MSO4IUIKGHP6FJ6D.anvil.app/Y6FBOUO22J75MCS2L75XZDSG)

Please bear in mind files will be saved to the data tables so please don't upload anything sensitive if using the example link.

**Reply by s-cork** [reply]
Wanna check this out - you'll need to adjust the url in the loader module to match the api endpoint url

https://anvil.works/build#clone:QDIQLVBJNZB6NDGB=2KAH3D4DSCVDNA4WJCYNRXHR|C6ZZPAPN4YYF5NVJ=|XCOXK7RMKUZ4GRXD=PTRRADDWEUMQQ2T3HGKA6NDN

It will clone a local copy of anvil_labs because I found a couple of bugs when playing with this

Let me know what you think
Just a proof of concept for now

**Comment by rhurlbatt** [comment]
!!!! Its sooooo freaking Awesome!!!!

Honestly Stu if you ever need a personal cheerleader - I am there, pompoms and all.

Love it - now, how the heck do I use it 😂
The only issue I picked up was that if the another worker is launched at the same time it errors out  but I will do some looking in to this.

So I am presuming we could use this to send anything JSONable right?
My idea is to create a module to take a basic JSONable data table row, store it in Indexed DB and then user the web worker to ship it off when internet connection is good.
At the endpoint this could then merge changes in to the live row. 

Thank you for checking this out Stu - even this is immensely helpful so far.

**Reply by s-cork** [reply]
yeah exactly - you could handle the api endpoint by creating a media object from the response body and then storing that in the table.
And yeah anything jsonable or just a bytes object can be sent with browser fetch api.


An alternative is simply to wrap the saving mechanism you currently have with: `with anvil.server.no_loading_indicator`.
The only advantage of the web worker here is that any new server calls from interaction with the app will be queued whilst the outstanding calls take place.
With the `no_loading_indicator` approach the a user can still interact with the app while the data is being saved to the table rows.
Might be worth exploring.

```python
from functools import wraps

def no_loading_indicator(fn):
    @wraps(fn)
    def wrapper(*args, **kws):
        with anvil.server.no_loading_indicator:
            return fn(*args, **kws)
    return wrapper

...

  @no_loading_indicator
  def write_from_indexed_db_to_datatable (self, **event_args):  
    self.files_remaining = len(self.media_store)
    ...
    
```


Re errors with multiple tasks being launched - You can create a new worker or just wait for a task to be completed before trying run a new one.
That should be as simple as checking whether `self.task is None` or not.

**Reply by rhurlbatt** [reply]
> An alternative is simply to wrap the saving mechanism you currently have with: with anvil.server.no_loading_indicator.

This was my first approach but since the app is so anvil.server.call() intensive you literally end up waiting for pages to load since they require data from the server (and there is no spinner to indicate something is going on).

>With the no_loading_indicator approach the a user can still interact with the app while the data is being saved

But they can't make server calls as they get queued.

>The only advantage of the web worker here is that any new server calls from interaction with the app will be queued whilst the outstanding calls take place.

As I understand it, the web worker works on a separate thread right? So the app just continue as per normal (presumably there is some minor impact to bandwidth or processor efficiency or something). In which case this web worker is ideal.

I am rebuilding an inspection app over the next few days - are you happy for me to try work this into it? 

Edit: I have just confirmed that with the `anvil.server.no_loading_indicator` approach, server calls are queued until the data is sent where as with web workers it is processed as per normal - excellent result!

**Reply by s-cork** [reply]
yeah - that's the caveat - the server calls are queued. So yeah you're right - in that case the web worker is probably the right approach here!

**Reply by s-cork** [reply]
> I am rebuilding an inspection app over the next few days - are you happy for me to try work this into it?

Yeah absolutely.
anvil-labs always comes with the warning that it's experimental so always worth keeping that in mind.

I've updated the module so you can use real anvil_labs rather than the local copy
Or just copy the `web_worker` module into your project, since it's entirely standalone

**Reply by rhurlbatt** [reply]
Our entire company/app is experimental so that fits perfectly hahahaha

Awesome - I will add it in and report back

Thanks again Stu!

**Comment by rhurlbatt** [comment]
So I have got this working up to the point of the HTTP endpoint, but I am struggling to convert the request body into a `BlobMedia` object to store it in the database.
This is the server function (pretty much as it was) but it raises an error - something to do with the way I am handling the Bytes.
```
@anvil.server.http_endpoint("/save_media")
def save_media(**params):
    print(params)
    req = anvil.server.request
    body = req.body.get_bytes()
    blob = anvil.BlobMedia(body, params['type'], params['name'])
    app_tables.media.add_row(file=blob)
```

prints:
```
{'name': 'download (3) (1).png', 'type': 'image/png'}
b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00\x00\x8f\x00\x00\x00\x8f\x08\x03\x00\x00\x00\xf4\xbdx\xa1\x00\x00\x00lPLTE\xff\xff\xff*\xb1\xeb\xff\xff\xfd*\xb0\xec\x00\xa8\xea\xdf\xf1\xfa\x18\xab\xe8\x00\xab\xea\x00\xa9\xe9\xdb\xf0\xfc\x1f\xaf\xea\x1b\xae\xeb\xe3\xf3\xfb\xd6\xed\xf9\xfb\xfe\xff\xf7\xfc\xfd\xed\xf7\xfck\xc3\xf0~\xc9\xef\x88\xcf\xf1P\xbb\xeeh\xc4\xee\x98\xd4\xf2\xcc\xe8\xf8[\xbf\xee\xc6\xe5\xf8\xbf\xe2\xf8\x00\xa4\xe8r\xc8\xf1;\xb5\xed\xb5\xe1\xf8\x8a\xcc\xf3\x9c\xd9\xf2\xa9\xda\xf6\xb7\xe3\xf4@\xb9\xe9\xea\xc1\xf6~\x00\x00\x06\xecIDATx\x9c\xed\\\x0bs\xa3*\x14V\xd0\x80\xa0\x11\x1fQ\x93\x06\x9b\xb4\xff\xff?^\xb4mb\x1e\xc0A1;s\'\xdf\xec\xcc\xcev\xad~\x1c\x0e\xe7\xadA\xf0\xc6\x1bo\xbc\xf1\xc6\xff\x0fh\x80\xfa\xeb\xf7_\xff\x96L0\xf2\xf9\xd7\x14\x06\xa0\x1fQlE\x92\xf5MY|\x87!\x0bCY\x17\x9f\xa7}\x96\x8a\xdf\x0b^\xc7t\x94J~(\x8b\x9aQ\xc29SP\x8c\x18\xe3\x9c\x10\x8a\xbf\x8fM\xf6s\xcd\xcb\x18\x89\xb6\x92\x1f\x84)\xa9\xe0\xf0\x0eX\x11\xe3\x94~\x9f\xb3\xed+\x98\xa8Uo\xb3R*\xa1\xdc\x13\xb9\x05#a\xdd\x88\x17\xa8Wz\x0e)\x7f\x90\xca3\xe0M\xb4\xeb\xb7\xeb2\x8a\x8f\xa1M2\x93\x9d\x0b9\x97g\xb1\x16\x17\x84\xe2#\xe5@2\x97}\xa3\xb2\xd9\x06\xab\x08)=\x12\x06\xda\xa8{!\x91s\xee\x9b\x8fZ`#\xa1\x1b\xf5\x00R\xb7^\t)6{I\xe6\xb2\x19@\x8f\xc2\xa79\xca\xabp\xb6p\x06`\xcc\xe4\xde\x1b\x9b \xae\xe9\x126#\x18\xad\x90\x1f\xbdF_\xc4\xf5T=\x95\x11\xa9c/;V.\xd2\x9c\t8o\x17K\x08\xe5E\xe4\x89\xce\xe0D\xceK\x95Zt\xbe\xa4\x13\x0e\xc6\x88\x9c\x97\xf1IC\x1f\xaa3\x01\xf9\\\xe2c\xe3\xda3\x1d\xa5D\x9f\xb3\xd9 \xe1[:\x03\xa2f\xa6\x80\x90\xe8V\xa0\xa3\xb6\xac\x99\xa9\xd4\xddf\r:!\xa6\x879\xd2A\xe5*\xd2\x19\xc0\xe7\xf8\x8e\x13\xc5\xce\xd1\x05\x14L\xb8\xe9\x90\xf24\xad{\xac\x03\x85\n\xf9k\xc7`\x1f\xe5r\xb5\xdd\x1a@*G\xcfq$\x8f\xa9\x8cG`\xb6w\xe2\xd3{\xf4\x12O\xc1B\x978\x1f\xad\xbaY#x\x05\x96\x0f\n\x8e\xb0\x0ck\x19\xa1\x16X\x10A([\x14\x9cB!\xc1*=?\x93p\x01\xe9a\xd2\t\x0e\xebk\xcf\x00V\x83T\x1a\xe5\xf5K\xe8\x84\xc0\xe0\x0c\x1d\x96\'\x130`\x0e\xe2S\xbfD{F>\x80p\x1a\xed_\xa3=\x03x\rH\xec\xd7\x0b3\x1eA\xecy\xbd\x88\xd67\x85\x17\xe0\xda*\x9efm\xcfu\xc3\x87\xc56\xf5q\xd3\xe6h\x1e.\xcfP\x1amd\x13\xb4`:D\xe1C\xc4s\x90\x16\x97\xa7\xd8\x02\xb3\xe9va\x03T\x0c|P@\xb3\x10\x94\x17>\x910i4\xba\xba.[\xf0\xbcY\xd2,\xb8\xca\x87\x9c\x8c7\xc9/\xc5\x03\x8c7\x06\xd0M4\x93\xca\xf8\x94\xee\xaa@\x85\xd1$^\xe3B\x8b\xa6-\x00\n\xfa\xab\x89c25];1\x86\xb0p\xc0\x15\xdb\\\x88\xc3\xb4\x84\xcdZ\xc3\xd5\xf9\xf4\xb4\xd3\xbd\xa7&\xcd_k%\xcf\xce\xcd\xee\x9b\xd1\x9bT\x81\x9a\xf6!\xbd\x89\xc4x\xef\xad\x90\x9d\xa7m%C>\xf4\x82\xee\xce\x05;\x1a\x1e\x91\xd1\ts\x1c\xd2/?|\xe2\xa6S\xe6\xeaI3h|\x8a\xe1\x17\xef\xd3\x1c\xfa\x99/\xda\xb1a\xa3\xb2\x86SS\xcf\xe3C\xff\xdbAu\xe7\xdb1\xe9\x8c\xf6\xcaFG\xad\xb0\xc0\x96\\%\xd2\xba0\x14t\xf7\xeb\xc0\x985\xbfw\x9e\xc3\xa7\x97t\xdc\x12\x13\x88\xa1\xd8\xf1$\xb1\xc0c\xf3a\x0e\x1f\x04\xaa\x85b\x93\xa1{\xfa\xfb,*f\xd5\xfa3P\xb5\x0f\xb3R{\x87\\\xa3v\x9c\xc9s\x8a\x9c\x1c\x16\n\x12\n\x0b\x15X\xa9\xbdi\xaa[\x10\xc6$<6N\x15\x00\x01L*1\xd3z0\x94\xe8n\x81\x87v6\x896\xbbs\x9b\xa5b\xfb\'\x02\xed\xc2\xd4\x7f\xec\xa0a\xb8\xe2\xa3Cb\xbb\x07#$\x94u]\x14E5\xf6\xb3u\x84\x10\xca\xc0I\x1c.\xf2\xd9|\xc2\xd1\x02\xa8?\x9c\x93\x88\x95\x99\x96\x8fr\xcc\xd0\xb4\x80uz>n\xb1<\t\x1b\x1d!\xa8\xf6x\xe5\xa3\xfc\x89&&A\t\xbc)\xe4\x91\x0f\xe6\xf2\xf9\x99C_\xf0\x1a\x00.t!=\x8a\x9d3w\x9a=\xbf\xd3\x19\xbe2\xfdy\x0f\x84s\xaaL4\xa5\x7f\x87z\xa8\xb2\xcf:\xfb\xe3^\xc6\xa4\xcf\x13p\x87\xe3n\xe2\x13`\xd7\r\xd3\xe4\xbbH\x80o\x84y\xa3\xd9-\x05\xd7\xd2\x0f+\x9e/\rM2,\x1b\x1fS\xb7\xc7\xb5\xd6\xf2\x91\xea<lK\xa0\xf60J\xf4|Nn\x07\x9e\xf7z\x1f\x06^\xda\x87!\x83o\x9dz\xdbQ\x83\xf4.,\x07\xb6^1\xd1\xd3\tb\x97&Jd*\x8f\xaaP\t6dap\xef\xca\x00\xc1K\xbd\x9c\xef\x8d\xc1\x19ByI\x00JM\x0c\xc7+\x80\x1e\x0b\x95\x9a\x15\xd6\x16\x1f\n\xf6\xa1\xf2\xf2\x16\x893\xe3\xb2N\xa0\x16\xae\xf2[\x07@\xce\x81\x02Q2K\xdc\xc1dj\xbaO\x0cQhB\xc6\xd0\xf5*\x9f\xeb \r\x9a\xfcx\xcc\xda\x0f\xdf\xd4\xd8\xf8d\xb5yYv\xf90\xd9<\x06\x08"k\xf7\x87C\x96\xa5\xdb\x07\xb1\xf5\xb54\xdc\x8d[\xa6\x03\xecf\x83W\xc1}\x99\xa1/jF(\xa5\x11\x91\xdd]\xd4\xaf\xae\xcc\x8f\x06\xa5T\xd9\xa9\x91Ok\x15\xd0\x8d\xbfQ\x8f\x13MD.U\x0b\x16r\xbaK\xefFW\x8f\xfa52i\xa9g\n\xab\x0b\xbb\xe5\x13\xb4\x0f\x83nD\xde\x95E\x0c|\x06a\x1b\x80\x00\x1b6\xe1\xa3\x1e\xdb\x7f\xf0\xdb\xd2\xe7\x90\x1a\xd1\x1e\xc8\x07sSulDb\xb3\xaa\xe4\x1a\xc4#]\x1cH2\xe0~I{\xc2k3\xd1\xbc\x8c\x93_\xc4-}n\xeeX\x9dL\xb0\xd3\xaa\x009Y\xe9 k\xa8\xa9\x92\xaf\x0b\xb4\xd7\x90\xc9EZ:\x90&<\xca-t<\x82\x94\x80\x02\x05zY\x87\x07\x13\x01(,!\xf1":!?Z\xc9(:[\xd4\xbc\xa8C\xc8m\xcd\xaf?F\xcei\xc6\x1c`\x8b-\x9c\xc0!\xbb\x9cO\'\x8cr`\xb9\rm\xbb\xf5{\xa8x\xe8\x8f@\xcb\x7f\x99\xbf\x99U-\x9cF\xb6\xaa\xd5w\x8c\xc0\x94\xf9w\xc7T\xb6\xb2\xee\x96\x99\xb2\xe4g\x84\xb2u\x05Dj\xa7\xd6\x91\x8a\xa7\xcek\x1a!\xc6R\xe7\x9a\xff\x8e\xac5\x0f\x89\x7fZ}\x8eXo\x0c\x08\x93\xca\xbd#\xa2\x02c\xe7\x97c\x80\xe0\xbb`V\xc7(Y\x87\x0e\xe9\xe6\xb6\xafZ`K\xc4\t\xbcv\x1c\xa6\x9d\xa0\xf7>T\x8b9l\x8aM\x83/\xf0\xdb^p:\x8bz\xc4{H\xcd\x04\xccf\x9c\xf1[\xd6\xb2\x8e\xa5\xbfI_I\xcc\xd3\x1a\x10\xf8|\xc9\x00\x14\xbf[\x91\x97^&\xd7\xf10\xef\xe0\x03*\x0b\xddxP"&\x13?o\xa4)\x19g\xcb\xa6\xd7\xc7\x9c\xbe\x1c\xaaF~\xc6A\x94\xb7\xf9d\x0b\x18\xe1\x90H\x9fo\x10\x0e}\xeeD\xce?\xf9<\xaa\x84\xff\x17\xac\xfbY\xaf4*\x9bCv.\xb1)\x14C%\x8c8\xcb\x083\xf2\xbd\x7f,*\xfa!\x84\xf2f\x18\x10\xb1\xd6\x94\xa7\xb2\xe9\xf6\xc1\x9a\x1f\x0e\x10}\xfd0\xe2\xa4\x83\xf2}\xbb\xc4\xfb\xab\xb9\xb7P\x9a\xddVadu\xb3\xc38CwN\xd7ys\xf9\x1e\xdb\xb6\xac\x19\xe1\xcaj?\x1a\xee\xe1\x07\x98\xff\x14~_\xf6}\x00\x14\xe4Y\x7fd\x94\x12\xf6\xf3\xb1\x82?\xa1`\xce\t\x8ddu\x88_\xfb\xad\x87\x9f\x8f\\\xe4\xd9\xa9<v\xb5\x0c\xd9f3\x9c\xbc\xba\xee\xca\xcf>\x0b~;\x08\xff\xe4\xd3\x13(\x17i\x9cdY\x96$\xb1\x10\xdaQ\x83\xd7\xf1\t~>\t\xf2\xf7i\x907\xdex\xe3\x8d7\xdex\x05\xfe\x03Xyd\xcc\xe5\xbf\t\xca\x00\x00\x00\x00IEND\xaeB`\x82'
TypeError: BlobMedia content must be a byte string.
at /downlink/anvil/__init__.py, line 82
called from InspectionsOTS.WebWorker, line 14
called from /downlink/anvil/_server.py, line 1461
```
I notice in the Chorme Dev tools that looking at the Indexed_DB entry the content is stored as a `content: Uint8Array()`.
I am pretty sure this is something simple I am missing but any help would be Greatly appreciated.

**Reply by s-cork** [reply]
I think you just need to change the call signature to BlobMedia. Content type is first no?

Or just use keyword arguments.

**Reply by rhurlbatt** [reply]
#facepalm
There is a reason you shouldn't code when tired! 

Thank you Stu its working now.

We have a little more to do but should be able to put this WebWorker through its paces in the next few days.

**Comment by Coastline23** [comment]
Hope it's ok to ask this here as it's pretty closely related to the discussion and work already done.

@s-cork how difficult would it be to use a [PWA Service worker background sync](https://developer.mozilla.org/en-US/docs/Web/API/Background_Synchronization_API) for this sort of update?

That way the update should be able to cope with a loss of internet and still update even if the app is closed/no longer in foreground or the device is off(asleep) once internet returns.

The App I'm developing requires a messaging function so the ability to enter a message when offline click send put the phone back in your pocket and have the message sent once the phone gets a connection is much preferable to only sending once you go back to the app (possibly hours or days later).

This would truly make the offline usability of Anvil awesome!!

**Comment by rhurlbatt** [comment]
@s-cork Are you able to help me adjust the code to send a dict. I have the media working but have not been able to send a dict via http.

```
# the Worker Function
def save_dicts(dicts_to_send):
  debugger
  worker.task_state["current"] = None
  worker.task_state["remaining"] = len(dicts_to_send)

  for i_uuid, i_dict in dicts_to_send.items():
    print (i_uuid)
    print (i_dict)
    if not i_dict.get('synced'):
      worker.task_state["current"] = f"Inspection {i_uuid}"
      
      worker.fetch(f'https://7BFWWU5AZVWM5IVB.anvil.app/4UGUTFZLNVQR5ZQ3TGT5HDZ5/_/api/save_inspection', {
        "method": 'POST',
        "headers": {"Content-Type": "application/json"},
        "body": i_dict
      });
      print (f"{i_uuid} uploaded")
      worker.task_state["remaining"] -= 1

# The HTTP func

@anvil.server.http_endpoint("/save_dicts")
def save_dicts(**params):
  print(params)
  req = anvil.server.request
  bb = req.body.get_bytes()   # Have also tried using body bytes
  
  print (bb) # prints [Object object] ?

  row= app_tables.table.get(i_uuid=bb['i_uuid'])

  if not row:
    row= app_tables.table.add_row(i_uuid=bb['i_uuid'])

    row.update(**bb)

  return "thanks"
```

I have tried to send it as a string which works but figured there was something I am missing in my understanding of the fetch api.
Do I need to convert the dict to soemthing JSONable before sending it?

**Reply by s-cork** [reply]
Yeah you'll just need to do something like `worker.JSON.stringify(idict)`. 

Or convert to json before you send it to the worker. 

If you choose application/json then the body must be json.

**Reply by rhurlbatt** [reply]
Nailed it thanks Stu .. I had tried a bunch of other approaches but did not know about the `worker.` option.
I used `worker.JSON.stringify(idict)` and added the `Content-Type` header to it and the dict comes out perfectly the other end.

**Comment by rhurlbatt** [comment]
It seems that this WebWorker code has now stopped working.
I don't see any PRs that would have changed anything in the WebWorker file.
Using the Clone link above to create a fresh version still doesn't seem to get it working.
The WebWorker loads the task but never completes a fetch and the number remaining doesn't change.
Have there been any changes here or in Anvil that would impact this?

Thanks in advance.

**Reply by s-cork** [reply]
Works for me.
no changes to the library
recloning works as expected - I switched to using the third party version of anvil_labs rather than my local copy
I also added a function, which allows you to test it in debug mode

Here's a slightly updated clone - https://anvil.works/build#clone:6KH7IIUS22JMNW54=664RQKGADXDEV6KC4XPXNMPR

You might want to add an error handler to the task in case there are errors that you're not seeing

```python
        self.task.on_error = self.error_handler

    def error_handler(self, error):
        raise error
```

**Reply by rhurlbatt** [reply]
Thanks Stu, This updated version works for me but still cloning the original and replacing the endpoint url still has issues.
I cannot see what is different or why this would be failing.

In the failing app's logs I see this though. Two HTTP calls, 1 second apart with the same uuid of the file that is being loaded.
Interestingly the session that started first has no events in it.
![image](https://user-images.githubusercontent.com/31876600/186812625-1498af31-51c0-4681-926a-28669fd8a593.png)

The updated version shows no HTTP sessions at all but completes correctly (3 different sessions shown - all completed correctly)
![image](https://user-images.githubusercontent.com/31876600/186812525-61428f74-e8f5-478a-845c-3f244daf95a5.png)

**Reply by s-cork** [reply]
The original version was only ever supposed to work in published version of the app (proof of concept).
If you turn on `enable_cors` in the `http_endpoint` decorator then I think it would have worked in debug sessions too.

The current version gets the api endpoint associated with the current session. So it's a bit cleaner. And should work in debug sessions.

For the logs - there's probably a same origin thing going on in the debug session.
I expect you'll get the HTTP logs in those browser sessions.

If debug sessions worked before and not now - then I'm not sure.
(I also can't remember if they did or not)
